src/io/trailer.cc

```diff
@@ -0,0 +1,156 @@
+// Copyright (C) 2019-2023 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License
+
+#include "io/trailer.h"
+
+#include <cstring>
+#include <fstream>
+
+namespace {
+static constexpr size_t kBlockSize = 4096;
+uint32_t
+CalculateCheckSum(const uint8_t* data, int64_t size) {
+    uint32_t checksum = 0;
+    for (auto i = 0; i < size; i++) {
+        checksum ^= data[i];  // xor
+    }
+    return checksum;
+}
+
+uint32_t
+GetFilesCheckSum(std::vector<std::string> files) {
+    uint32_t checksum = 0;
+    auto buffer = std::shared_ptr<uint8_t[]>(new uint8_t[kBlockSize]);
+    for (auto& file_name : files) {
+        std::ifstream reader(file_name.c_str(), std::ios::binary);
+        if (!reader) {
+            LOG_KNOWHERE_WARNING_ << file_name << "not exist, skip calculate check sum for this file.";
+            continue;
+        }
+        while (reader.read((char*)buffer.get(), kBlockSize)) {
+            std::streamsize read_size = reader.gcount();
+            checksum ^= CalculateCheckSum(buffer.get(), read_size);
+        }
+    }
+    return checksum;
+}
+}  // namespace
+
+namespace knowhere {
+void
+AddTrailerForMemoryIO(MemoryIOWriter& writer, const std::string& name, const Version& version) {
+    auto trailer_ptr = std::make_unique<Trailer>();
+    auto size = writer.tellg();
+    trailer_ptr->SetIndexBinarySize(size);
+    trailer_ptr->SetCheckSum(CalculateCheckSum(writer.data(), size));
+    trailer_ptr->SetVersion(version.VersionNumber());
+    trailer_ptr->SetIndexName(name);
+    writer.write(trailer_ptr->bytes, KNOWHERE_TRAILER_SIZE);
+}
+
+bool
+CheckTrailerForMemoryIO(MemoryIOReader& reader, const std::string& name) {
+    uint64_t bin_size = TRAILER_OFFSET(reader.size());
+    if (bin_size < 0) {
+        LOG_KNOWHERE_WARNING_ << "The binary is too small and assume no Trailer, pass Trailer check.";
+        return true;
+    }
+
+    auto trailer_ptr = std::make_unique<Trailer>();
+    auto pre_rp = reader.tellg();
+    reader.seekg(bin_size);
+    reader.read(trailer_ptr.get(), KNOWHERE_TRAILER_SIZE);
+    reader.seekg(pre_rp);
+
+    if (!trailer_ptr->TrailerValidCheck()) {
```

Comment:
what if this file has no trailer but these 4 bytes happens to be the same as TRAILER_VALID_FLAG?

is it possible to use version number to decide whether a index file should include trailer?

Reply:
I added flags at the head and tail of the trailer for checking.